### PR TITLE
[CPU] FusedAdam and CPU training support

### PR DESCRIPTION
**Pull request by delock**
This PR implements FusedAdam and also enables CPU training with ZeRO stage 0-3.  Note FP16 data type are not supported for CPU training.  BF16 is the preferred datatype.

For ZeRO stage 0, this PR depends on #3842.

Whats in this PR:
1. FusedAdam kernel, the implementation is from CPU Adam kernel and had been modified to suite the case.
2. Fix bugs in CPU Accelerator interface implementation which had been exposed by training workload.
3. Handling of synchronized accelerator in DeepSpeed runtime.

This PR had been verified on CIFAR10 in DeepSpeedExamples, A PR had been submitted to support BF16 data type for CIFAR10 https://github.com/microsoft/DeepSpeedExamples/pull/651 .  Note for Zero0/1, there is an accuracy bug in DeepSpeed supporting BF16 training (https://github.com/microsoft/DeepSpeed/issues/3979), for Zero2/3, accuracy is correct.

### CODE REVIEW

File: accelerator/cpu_accelerator.py [hunk 1]
```diff
@@ -139,7 +139,7 @@ def reset_max_memory_cached(self, device_index=None):
         return
 
     def memory_stats(self, device_index=None):
-        return self.get_rss()
+        return None
```
Comment (tjruwase) [inline accelerator/cpu_accelerator.py:142]: Why return `None`? Is there some other way to observe memory statistics of cpu_accelerator?

File: csrc/cpu/adam/cpu_adam.h [hunk 2]
```diff
@@ -0,0 +1,148 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0
+
+// DeepSpeed Team
+
+#pragma once
+
+#define NOMINMAX  // Windows idiosyncrasy
+                  // https://stackoverflow.com/questions/4913922/possible-problems-with-nominmax-on-visual-c
+
```
Comment (tjruwase) [inline csrc/cpu/adam/cpu_adam.h:10]: This file has some duplication with `csrc/include/cpu_adam.h`. Do you plan to merge them?

File: csrc/cpu/adam/simd.h [hunk 3]
```diff
@@ -0,0 +1,140 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0
+
+// DeepSpeed Team
+
+#pragma once
```
Comment (tjruwase) [inline csrc/cpu/adam/simd.h:6]: This is also similar to `csrc/includes/simd.h`. It will get good to merge to avoid duplicate files in the repo.

File: accelerator/cpu_accelerator.py (hunk 1 above)
Comment (delock) [reply accelerator/cpu_accelerator.py:142]: From the usage of memory_stats caller is expecting a dict or None.   The attribute DeepSpeed is using is `num_alloc_retries` which is GPU specific.
https://github.com/microsoft/DeepSpeed/blob/fb9aebbf254d6d4535e42df75cb08b7c1dbd03f2/deepspeed/runtime/zero/stage3.py#L1895

Any suggestions?  Besides `None`, alternative options is return a dict with attribute `num_alloc_retries` set to 0 and emulate other attributes in the future when needed.

File: csrc/cpu/adam/cpu_adam.h (hunk 2 above)
Comment (delock) [reply csrc/cpu/adam/cpu_adam.h:10]: the CUDA version of cpu_adam.h provide Adam in the form of optimizer object.  We changed it to stateless function call to fit the scenario of FusedAdam.  Besides the CUDA version contains CUDA specific code, so we think its better to make a copy of cpu_adam.h to decouple.

File: csrc/cpu/adam/simd.h (hunk 3 above)
Comment (delock) [reply csrc/cpu/adam/simd.h:6]: This file is identical to `csrc/includes/simd.h`, let me see if the duplication could be removed.

File: accelerator/cpu_accelerator.py (hunk 1 above)
Comment (tjruwase) [reply accelerator/cpu_accelerator.py:142]: Thanks for pointing this out. Here are my thoughts. 
1. We should return `dict` of interesting memory statistics. I think it is okay if keys are different across accelerators.
2. We should change DeepSpeed code to first check `num_alloc_retries` in the returned dict before executing any logic.

What do you think?

Comment (delock) [reply accelerator/cpu_accelerator.py]: Updated memory_stats to report `allocated_bytes.all.current` and `allocated_bytes.all.peak`, also check whether key `num_alloc_retries` exist first in use point.

File: csrc/cpu/adam/cpu_adam.h (hunk 2 above)
Comment (tjruwase) [reply csrc/cpu/adam/cpu_adam.h:10]: I understand. In the long run, it would be nice to remove duplicate code by having the CUDA cpu_adam inherit from the basic cpu_adam. But that will be a future PR.

**Comment by delock** [comment]
@tjruwase note we also have a PR in DeepSpeedExamples to demostrate CIFAR10 training with CPU.  The PR also fix test broken for pytorch 2.0 because of API change.
https://github.com/microsoft/DeepSpeedExamples/pull/651

File: csrc/cpu/adam/cpu_adam.h (hunk 2 above)
Comment (tjruwase) [reply csrc/cpu/adam/cpu_adam.h:10]: @delock, apologies for revisiting this issue. But I was reminded that we had made cpu_adam runnable in CPU only environment with PR #2507. Moreover, we created this [unit test](https://github.com/microsoft/DeepSpeed/blob/23a11a39510e2aefb48236e3d2672a7dcbfc42a3/tests/unit/ops/adam/test_cpu_adam.py#L90) for this scenario since `torch.optim.AdamW` can also work in cpu only environment. 

So I am unsure that the changes in `csrc/cpu` are actually needed. Did you try to use existing for to support CPU training? Thanks!

**Comment by delock** [comment]
Also added CPUAdamBuilder for CPU accelerator, so HelloDeepSpeed which use cpu offload can run.

**Review by tjruwase — APPROVED** [review]

File: csrc/cpu/adam/cpu_adam.h (hunk 2 above)
Comment (tjruwase) [reply csrc/cpu/adam/cpu_adam.h]: Awesome! Thanks!
